akka-projection-core/src/main/scala/akka/projection/TelemetrySpi.scala

```diff
@@ -0,0 +1,99 @@
+/*
+ * Copyright (C) 2020 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.projection
+
+import scala.collection.immutable
+
+import akka.actor.typed.ActorSystem
+import akka.annotation.InternalApi
+import akka.annotation.InternalStableApi
+
+/**
+ * Service Provider Interface (SPI) for collecting metrics from projections.
+ *
+ * Implementations must include a single constructor with two arguments: [[ProjectionId]]
+ * and [[ActorSystem[_]].  To setup your implementation, add a setting on your `application.conf`:
+ *
+ * {{{
+ * akka.projection.telemetry.fqcn = com.example.MyMetrics
+ * }}}
+ */
+trait Telemetry {
```

Comment:
This change looks good!

Probably it's not needed now but if we could create a custom context at the moment when the event has been received and then propagate it into the other handlers it would provide a more flexible SPI . Otherwise we would need to change this SPI if for example we would need to produce traces for read-side.

Reply:
I've looked into it and I think we can introduce the improvement here to allow externally-created contexts to propagate.

Reply:
raised https://github.com/akka/akka-projection/issues/276